### PR TITLE
feat: allow arbitrary extensions on nexus types

### LINKED ISSUE #683: How to set extensions on a type?
I found out that it is possible to extend the extensions on a field using the queryComplexityPlugin as a blueprint.
But now I would like to do the same for a type as well.

For my particular use case I am trying to use JoinMonster with @nexus/schema. For this to work JoinMonster (starting with version 3.0.0, see https://github.com/join-monster/join-monster/blob/master/docs/CHANGELOG.md#v300) requires to use the GraphQL extensions property introduced in GraphQL v15.

As a simple solution I tried to extend the queryComplexityPlugin example and in ```onCreateFieldResolver(config)``` run the following code to update the type extensions config:

```
const joinMonsterConfigOnType =
    config.parentTypeConfig.extensions?.nexus?.config.joinMonster;

if (joinMonsterConfigOnType != null) {
    config.parentTypeConfig.extensions = {
        ...config.parentTypeConfig.extensions,
        joinMonster: joinMonsterConfigOnType,
    } as any;
}
```

But it looks like Nexus is overwriting the type's extensions later on with its own. Could be this location here: https://github.com/graphql-nexus/schema/blob/develop/src/builder.ts#L963

Any help would be much appreciated.

### PR DESCRIPTION
**Pull request by tgriesser**
Closes #683

### CODE REVIEW

**Comment by jasonkuhrt** [comment]
pr release made:

```json5
 { distTag: 'pr.684', version: '0.0.0-pr.684.1.187f285' }
```

**Comment by benjixx** [comment]
I just discovered an edge case which seems to be not working yet.

I have an interface A and some objectType B that implements interface A. I would like to modify the implemented field that is part of interface A (via modify method) within objectType B and also set the extension for the field there.

**Review by jasonkuhrt — APPROVED** [review]
LGTM, maybe use the feature in the integration tests to "cover" this?

**Comment by tgriesser** [comment]
> I have an interface A and some objectType B that implements interface A. I would like to modify the implemented field that is part of interface A (via modify method) within objectType B and also set the extension for the field there.

Cool, will look into supporting this

**Comment by jasonkuhrt** [comment]
@benjixx added a failing test for that case.

File: tests/interfaceType.spec.ts [hunk 1]
```diff
@@ -196,3 +196,49 @@ describe('interfaceType', () => {
...
+    ],
+    outputs: false,
+    shouldGenerateArtifacts: false,
+  })
+  const A = schema.getType('A') as GraphQLObjectType
+  expect(A.getFields().foo.extensions).toMatchInlineSnapshot(`
+    Object {
+      "deeply": Object {
+        "foo1": true,
+        "foo2": true,
```
Comment (jasonkuhrt) [inline tests/interfaceType.spec.ts:240]: @tgriesser do you agree extension modifications should deeply merge?

File: tests/integrations/kitchenSink/__typegen.ts [hunk 2]
```diff
@@ -80,10 +80,7 @@ export interface NexusGenObjects {
     // root type
     hello?: string | null // String
   }
-  OfI2: {
-    // root type
-    hello?: string | null // String
```
Comment (jasonkuhrt) [inline tests/integrations/kitchenSink/__typegen.ts:85]: @tgriesser Looks like a bug to me. If user uses `t.modify` only to augment `extensions` then why should the default backing types be altered?

File: tests/interfaceType.spec.ts (hunk 1 above)
Comment (tgriesser) [reply tests/interfaceType.spec.ts:240]: I think they should probably stay independent until we have a use case for them to be merged.

File: tests/integrations/kitchenSink/__typegen.ts (hunk 2 above)
Comment (tgriesser) [reply tests/integrations/kitchenSink/__typegen.ts]: Yep, this is all pretty messy right now. Got it working temporarily, but will try and refactor this as a part of fixing #673

**Pull request merged**